src/p11_wrap.cpp

```diff
@@ -0,0 +1,274 @@
+{
+    if (ctx == nullptr) {
+        throw P11Exception("NULL passed for context when releaseing slots");
+    }
+
+    if (slotInfo._slots == nullptr) {
+        throw P11Exception("NULL list when releaseing slots");
+    }
+
+    lib::LibP11::P11_PKCS11_release_all_slots(ctx, slotInfo._slots, slotInfo._numSlots);
+
+    // Finally, clear the object to remove dangling information.
+    slotInfo._slots = nullptr;
+    slotInfo._numSlots = 0;
+}
+
+P11_PKCS11_SLOT_PTR _PKCS11_find_token(PKCS11_CTX *ctx, P11_SlotInfo &slotInfo)
+{
+    if (ctx == nullptr) {
+        throw P11Exception("NULL passed for ctx");
+    }
+
+    return P11CallPtr::callChecked(
+            lib::LibP11::P11_PKCS11_find_token, ctx, slotInfo._slots, slotInfo._numSlots);
+}
+
+P11_PKCS11_TOKEN_PTR _PKCS11_getTokenFromSlot(PKCS11_SLOT *slot)
+{
+    if (slot == nullptr) {
+        throw P11Exception("Cannot retrieve token from NULL slot pointer");
+    }
+
+    // Simply obtain the token via the slot's data structure.
+    P11_PKCS11_TOKEN_PTR token = slot->token;
+    if (token == nullptr) {
+        throw P11Exception("Cannot get token. Slot not associated with token");
+    }
+
+    return token;
+}
+
+void _PKCS11_open_session(PKCS11_SLOT *slot, int rw)
+{
+    if (slot == nullptr) {
+        throw P11Exception("Cannot open session from NULL slot pointer");
+    }
+
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_open_session, slot, rw /* Denotes read/write mode. */);
+}
+
+void _PKCS11_login(PKCS11_SLOT *slot, std::string pin)
+{
+    if (slot == nullptr) {
+        throw P11Exception("Cannot login from NULL slot pointer");
+    }
+
+    if (_PKCS11_is_logged_in(slot)) {
+        throw P11Exception("Mismatch detected; user already logged in");
+    }
+
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_login, slot, 0 /* Always login as normal user. */, pin.c_str());
+}
+
+void _PKCS11_logout(PKCS11_SLOT *slot)
+{
+    if (slot == nullptr) {
+        throw P11Exception("Cannot logout from NULL slot pointer");
+    }
+
+    if (!_PKCS11_is_logged_in(slot)) {
+        throw P11Exception("Mismatch detected; user not logged in to perform logout");
+    }
+
+    P11IsNonNegative::callChecked(lib::LibP11::P11_PKCS11_logout, slot);
+}
+
+bool _PKCS11_is_logged_in(PKCS11_SLOT *slot)
+{
+    if (slot == nullptr) {
+        throw P11Exception("Cannot check login from NULL slot pointer");
+    }
+
+    int result;
+    P11IsNonNegative::callChecked(
+            lib::LibP11::P11_PKCS11_is_logged_in, slot, 0 /* Check normal user. */, &result);
+
+    // Result is 1 if logged in.
+    return result == 1;
+}
+
+void _PKCS11_store_private_key(PKCS11_TOKEN *token, EVP_PKEY *pk, std::string label, std::string id)
+{
+    std::string parsedStrID = parseID(id);
+    unsigned char *parsedID =
+            reinterpret_cast<unsigned char *>(const_cast<char *>(parsedStrID.c_str()));
```

Comment:
Every const case please with comment and statement that you checked that this is just due to bad const-correctness in upstream API.